scalanet/src/io/iohk/scalanet/peergroup/UDPPeerGroup.scala

```diff
@@ -98,31 +100,24 @@ class UDPPeerGroup[M](val config: Config, cleanupScheduler: Scheduler = Schedule
           .addLast(new channel.ChannelInboundHandlerAdapter() {
             override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
               val datagram = msg.asInstanceOf[DatagramPacket]
+              val remoteAddress = datagram.sender()
+              val localAddress = datagram.recipient()
+              val udpChannelId = UdpChannelId(ctx.channel().id(), remoteAddress, localAddress)
               try {
-                val remoteAddress = datagram.sender()
-                val localAddress = datagram.recipient()
                 log.info(s"Client channel read message with remote $remoteAddress and local $localAddress")
-
-                val channelId = (remoteAddress, localAddress)
-
-                val channel = activeChannels.get(channelId)
-
-                if (channel == null || !channel.isOpen) {
-                  // It should never happen as closing client channel closes underlying netty channel, so channelRead should
-                  // not execute on this channel
-                  throw new IllegalStateException(s"Missing channel instance for channelId $channelId")
-                } else {
-                  handleIncomingMessage(channel, datagram)
-                }
+                Option(activeChannels.get(udpChannelId)).foreach(handleIncomingMessage(_, datagram))
               } catch {
-                case NonFatal(e) => handleError(ctx, e)
+                case NonFatal(e) => handleError(udpChannelId, e)
               } finally {
                 datagram.content().release()
               }
             }
 
             override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+              val channelId = ctx.channel().id()
+              val localAddress = ctx.channel().localAddress().asInstanceOf[InetSocketAddress]
               val remoteAddress = ctx.channel.remoteAddress().asInstanceOf[InetSocketAddress]
```

Comment:
There's no avoiding of this asInstanceOf due to ChannelHandlerContext being something from netty, right?

Reply:
unfortunately yes. It even stated in docs that this is the way of handiling it:
```
The returned
     * {@link SocketAddress} is supposed to be down-cast into more concrete
     * type such as {@link InetSocketAddress}
```